syz-verifier/monitoring_api.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"encoding/json"
+	"net/http"
+	"time"
+)
+
+// Monitor provides http based data for the syz-verifier monitoring.
+type Monitor struct {
+	externalStats *Stats
+}
+
+// MakeMonitor creates the Monitor instance.
+func MakeMonitor() *Monitor {
+	return &Monitor{}
+}
+
+// InitHTTPHandlers initializes the API routing.
+func (monitor *Monitor) InitHTTPHandlers() {
```

Comment:
I think it's better to execute this as part of MakeMonitor or ListenAndServe. Currently the use of monitor is more complex then it needs to be (4 calls that must be done in a particular order, and not making any one of them is a programming error (e.g. no SetStatsTracking, or no InitHTTPHandlers). This object is not useful if InitHTTPHandlers is not called.

Reply:
Hmm... The logic was "create_monitor -> set_tracking_objects -> init_handlers->listen_and_serve".
Remove "init_handlers" step sounds good here,
Will support the "create_monitor -> listen_and_serve -> add_tracking_object -> add_tracking_object -> ..." then.

Reply:
I think it's better to do:
"create_monitor -> add_tracking_object -> add_tracking_object -> ... -> listen_and_serve"
This ways we avoid synchronization issues, exposing incomplete info to users and it just looks logically more reasonable.

Reply:
We can start monitor before VMs.
Later, when some client app is executed in VM env., we'll have the object to monitor.
In this case, Monitor() shouldn't be aware of VMs and environment - it'll get new objects dynamically.

Guess, there is a diff in our understanding of the use-cases. Requirements matters! )

Reply:
Guess I need to see the final use of the object. Let's go with your proposal for now.